base/reduce.jl

```diff
@@ -362,10 +367,95 @@ julia> reduce(*, [2; 3; 4]; init=-1)
 -24
 ```
 """
-reduce(op, itr; kw...) = mapreduce(identity, op, itr; kw...)
+function reduce(op, itr::T; kw...) where T
+    # Redispatch, adding traits
+    reduce(op, itr, eltype_or_default_eltype(itr), IteratorSize(T); kw...)
+end
+
+function reduce(op, itr, et, isize; kw...)
+    # Fallback: if nothing interesting is being done with the traits
+    # or the operation
+    return mapreduce(identity, op, itr; kw...)
+end
 
 reduce(op, a::Number) = a  # Do we want this?
 
+##### Operation specific reduce optimisations
+
+## vcat
+
+function reduce(::typeof(vcat), xs, T::Type{<:AbstractVector}, isize)
+    x_state = iterate(xs)
+    x_state === nothing && return T() # New empty instance
+    x1, state = x_state
+
+    ret = copy(x1)  # this is **Not** going to work for StaticArrays
+
+    if !(isize isa SizeUnknown)
+        # Assume first element has representitive size, unless that would make this too large
+        SIZEHINT_CAP = 10^6
+        sizehint!(ret, min(SIZEHINT_CAP, length(xs)*length(x1)))
+    end
+
+    x_state = iterate(xs, state)
+    while(x_state !== nothing)
+        x, state =  x_state
+        append!(ret, x)
+        x_state = iterate(xs, state)
+    end
+    return ret
+end
+
+## hcat
+
+function reduce(::typeof(hcat), xs, T::Type{<:AbstractVector}, isize::SizeUnknown)
+    x_state = iterate(xs)
+    x_state === nothing && return et() # New empty instance
+    x1, state = x_state
+
+    dim1_size = length(x1)
+    dim2_size = 1
+    ret_vec = copy(x1)  # this is **Not** going to work for StaticArrays
+
+    x_state = iterate(xs, state)
+
+    while(x_state !== nothing)
+        x, state =  x_state
+        append!(ret_vec, x)
+        dim2_size += 1
+        x_state = iterate(xs, state)
+    end
+
+    # Reshape will throw errors if anything was the wrong size
+    return reshape(ret_vec, (dim1_size, dim2_size))
+end
+
+function reduce(::typeof(hcat), xs, T::Type{<:AbstractVector}, isize)
+    # Size is known
+    x_state = iterate(xs)
+    x_state === nothing && return T() # New empty instance
+    x1, state = x_state
+
+    dim1_size = size(x1,1)
+    dim2_size = length(xs)
+
+    ret = similar(x1, eltype(x1), (dim1_size, dim2_size))
+    copyto!(ret, 1, x1, 1)
+
+    x_state = iterate(xs, state)
+    offset =  length(x1)+1
+    while(x_state !== nothing)
+        x, state =  x_state
+        length(x)==dim1_size || throw(DimensionMismatch("hcat"))
+        copyto!(ret, offset, x, 1)
```

Comment:
How do we know this will fit into `ret`?

```
julia> reduce(hcat, (UInt8[1,2], [1000, 5000]))
ERROR: InexactError: trunc(UInt8, 1000)
```

Reply:
we don't this function needs to be tightenned to 
```
T::Type{<:AbstractVector{S}}
```

And things that fell back though `@default_eltype` need to be banned from using it.

And then a nother (marginally slower, but benchmarking will tell)
version needs to be created that can deal with hetrogenous containers.
I would say that could be in another PR and I just want to common case covered here,
but actually i would be sad if can't get speedup for generators.

Reply:
> I would say that could be in another PR and I just want to common case covered here,

I don't understand, this PR breaks the use cases I linked so how can it be in another PR?

Reply:
Sorry I was unclear.
I mean definitately fixing that issue.
But the question of handling things like
```
(i for i in ([0x1, 0x2], [1,2]))
```
does not have to be handled, in this PR (I think it should, if it doesn't add undue complexity).
If we just tighten the type signature,
from `where T<:AbstractVector` to `where T<:AbstractVector{S} where S`.
That would solve your case by causing it to fallback to the current slow `reduce` methods.

Thus working out how to actually efficently handle hetrogeneous types could be in a PR;
if it was going to be hard.
But I don't think it will so it can be in this one.